src/frequenz/sdk/timeseries/_formula_engine/_formula_engine_pool.py

```diff
@@ -0,0 +1,116 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+"""A formula pool for helping with tracking running formula engines."""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Type
+
+from frequenz.channels import Sender
+
+from ...actor import ChannelRegistry, ComponentMetricRequest
+from ...microgrid.component import ComponentMetricId
+from ._formula_generators._formula_generator import (
+    FormulaGenerator,
+    FormulaGeneratorConfig,
+)
+from ._resampled_formula_builder import ResampledFormulaBuilder
+
+if TYPE_CHECKING:
+    from .._formula_engine import (
+        FormulaReceiver,
+        FormulaReceiver3Phase,
+        _GenericEngine,
+        _GenericFormulaReceiver,
+    )
+
+
+class FormulaEnginePool:
+    """Creates and owns formula engines from string formulas, or formula generators.
+
+    If an engine already exist with a given name, it is reused instead.
+    """
+
+    def __init__(
+        self,
+        namespace: str,
+        channel_registry: ChannelRegistry,
+        resampler_subscription_sender: Sender[ComponentMetricRequest],
+    ) -> None:
+        """Create a new instance.
+
+        Args:
+            namespace: namespace to use with the data pipeline.
+            channel_registry: A channel registry instance shared with the resampling
+                actor.
+            resampler_subscription_sender: A sender for sending metric requests to the
+                resampling actor.
+        """
+        self._namespace = namespace
+        self._channel_registry = channel_registry
+        self._resampler_subscription_sender = resampler_subscription_sender
+        self._engines: dict[str, "FormulaReceiver|FormulaReceiver3Phase"] = {}
```

Comment:
why some attributes of the class add type hints and some other don't? I'm asking here as I didn't want to interrupt you during the meeting/presentation

Reply:
maybe because the attributes set from the input parameters don't need it?